letsencrypt/renewer.py

```diff
@@ -0,0 +1,117 @@
+from letsencrypt import client
+from letsencrypt import crypto_util
+from letsencrypt import notify
+from letsencrypt import storage
+from letsencrypt.plugins import disco as plugins_disco
+
+DEFAULTS = configobj.ConfigObj("renewal.conf")
+DEFAULTS["renewal_configs_dir"] = "/tmp/etc/letsencrypt/configs"
+DEFAULTS["official_archive_dir"] = "/tmp/etc/letsencrypt/archive"
+DEFAULTS["live_dir"] = "/tmp/etc/letsencrypt/live"
+
+class AttrDict(dict):
+    """A trick to allow accessing dictionary keys as object
+    attributes."""
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self
+
+def renew(cert, old_version):
+    """Perform automated renewal of the referenced cert, if possible."""
+    # TODO: handle partial success
+    # TODO: handle obligatory key rotation vs. optional key rotation vs.
+    #       requested key rotation
+    if not cert.configfile.has_key("renewalparams"):
+        # TODO: notify user?
+        return False
+    renewalparams = cert.configfile["renewalparams"]
+    if not renewalparams.has_key("authenticator"):
+        # TODO: notify user?
+        return False
+    # Instantiate the appropriate authenticator
+    plugins = plugins_disco.PluginsRegistry.find_all()
+    try:
+        config = configuration.NamespaceConfig(AttrDict(renewalparams))
+        # XXX: this loses type data (for example, the fact that key_size
+        #      was an int, not a str)
+        config.rsa_key_size = int(config.rsa_key_size)
+        authenticator = plugins[renewalparams["authenticator"]]
+        authenticator = authenticator.init(config)
+    except KeyError:
+        # TODO: Notify user? (authenticator could not be found)
+        return False
+
+    authenticator.prepare()
+    account = client.determine_account(config)
+    # TODO: are there other ways to get the right account object, e.g.
+    #       based on the email parameter that might be present in
+    #       renewalparams?
+
+    our_client = client.Client(config, account, authenticator, None)
+    with open(cert.version("cert", old_version)) as f:
+        sans = crypto_util.get_sans_from_cert(f.read())
+    new_cert, new_key, new_chain = our_client.obtain_certificate(sans)
+    if new_cert and new_key and new_chain:
+        # XXX: Assumes that there was no key change.  We need logic
+        #      for figuring out whether there was or not.  Probably
+        #      best is to have obtain_certificate return None for
+        #      new_key if the old key is to be used (since save_successor
+        #      already understands this distinction!)
+        return cert.save_successor(old_version, new_cert, new_key, new_chain)
+    #    TODO: Notify results
+    else:
+    #    TODO: Notify negative results
+        return False
+    # TODO: Consider the case where the renewal was partially successful
+    #       (where fewer than all names were renewed)
+
+def main(config=DEFAULTS):
```

Comment:
please don't set function argument default values to mutable data structures! http://effbot.org/zone/default-values.htm